### PR TITLE
Bug 4829: IPC shared memory leaks when disker queue overflows

### PR DESCRIPTION
**Pull request by craiggowing**
The fixed leak was accompanied by these cache.log errors:

    ERROR: worker I/O push queue for ... overflow: ...

I/O queue overflows during disk read requests log the same error but do
not leak memory. Repeated overflows during disk write requests could
eventually exhaust IPC shared memory:

    ERROR: ... exception: run out of shared memory pages for IPC I/O

With IPC memory exhausted due to leaks, rock disk I/O stops forever.

### CODE REVIEW

File: src/DiskIO/IpcIo/IpcIoFile.cc [hunk 1]
```diff
@@ -372,6 +372,8 @@ IpcIoFile::push(IpcIoPendingRequest *const pending)
                dbName << " overflow: " <<
                SipcIo(KidIdentifier, ipcIo, diskId)); // TODO: report queue len
         // TODO: grow queue size
+        if (!pending->readRequest && ipcIo.page)
```
Comment (rousskov) [inline src/DiskIO/IpcIo/IpcIoFile.cc]: Please either do not check `pending->readRequest` or explain why this check is necessary. Please note that I know that the current code does not pre-allocate ipcIo.page for reading requests, but that fact does not matter here.

File: src/DiskIO/IpcIo/IpcIoFile.cc [hunk 2]
```diff
@@ -372,6 +372,8 @@ IpcIoFile::push(IpcIoPendingRequest *const pending)
...
+            Ipc::Mem::PutPage(ipcIo.page); // Free the page if allocated
```
Comment (rousskov) [inline src/DiskIO/IpcIo/IpcIoFile.cc]: Please remove the comment -- the code speaks (and speaks better) for itself in this case.

**Comment by rousskov** [comment]
I adjusted PR description (and future master commit message) to fix formatting, detail the symptoms, and make the text more suitable for a future [automated](https://github.com/measurement-factory/anubis#readme) commit. Please polish further as needed.

**Comment by craiggowing** [comment]
@rousskov - thanks for the feedback, I've updated the branch accordingly.

**Comment by rousskov** [comment]
OK to test

**Comment by yadij** [comment]
Seeing as Anubis seems to have halted just short of the final commit for several days I am merging this one myself now.

**Comment by rousskov** [comment]
> Seeing as Anubis seems to have halted just short of the final commit for several days 

@yadij, fully automated commits are not enabled yet. A human authorization is currently required. I was waiting for kinkie to create an account to authorize and monitor this PR commit. There are currently no other commits ready, so I will have to wait again now.

@eduard-bagdasaryan, please note that the M-failed-other label is wrong in this case. There may be a race condition with the merging/closing PR. Please investigate and propose a fix.

**Pull request merged**